llvm/include/llvm/CodeGen/MachinePassManager.h

```diff
@@ -129,16 +142,6 @@ class FunctionAnalysisManagerMachineFunctionProxy
       Arg.FAM = nullptr;
     }
 
-    ~Result() {
-      // FAM is cleared in a moved from state where there is nothing to do.
-      if (!FAM)
-        return;
-
-      // Clear out the analysis manager if we're being destroyed -- it means we
-      // didn't even see an invalidate call when we got invalidated.
-      FAM->clear();
-    }
-
```

Comment:
This causes double free.

Reply:
do you know why? I know that the analysis managers have to be declared in a specific order due to these sorts of things. does this still occur even with the following order:
```
  MachineFunctionAnalysisManager MFAM;
  LoopAnalysisManager LAM;
  FunctionAnalysisManager FAM;
  CGSCCAnalysisManager CGAM;
  ModuleAnalysisManager MAM;
```
?

I haven't thought too hard about whether or not this is actually required (meaning maybe we can't just remove this code for correctness)...

Reply:
It would be the circular reference between `MachineFunctionAnalysis` and `FunctionAnalysis` in `FunctionAnalysisManagerMachineFunctionProxy` and `MachineFunctionAnalysisManagerFunctionProxy`.

Consider:
```c++
MachineFunctionAnalysisManager MFAM;
// ...
FunctionAnalysisManager FAM;
// Fetch some analysis proxies

FAM.clear();
MFAM.clear();
```
`FAM.clear();` will clear `MachineFunctionAnalysisManagerFunctionProxy` -> `MachineFunctionAnalysisManager` -> `FunctionAnalysisManagerMachineFunctionProxy` -> `FunctionAnalysisManager`, i.e. clear `FAM` will try to clear itself again indirectly. This is also true for `MachineFunctionAnalysisManager`.

Reply:
ah this code was copied from `InnerAnalysisManagerProxy` where an outer IR analysis manager can clear an inner IR analysis manager, which makes sense, but it doesn't make sense for a MachineFunctionAnalysisManager to clear a FunctionAnalysisManager. my bad for this code